dvc/tree/gdrive.py

```diff
@@ -510,35 +512,88 @@ def exists(self, path_info, use_dvcignore=True):
         else:
             return True
 
-    def _list_paths(self, prefix=None):
-        if not self._ids_cache["ids"]:
-            return
+    def _gdrive_list_ids(self, query_ids):
+        query = " or ".join(
+            f"'{query_id}' in parents" for query_id in query_ids
+        )
+        query = f"({query}) and trashed=false"
+        return self._gdrive_list(query)
+
+    def _ls_recursive(self, path_info, detail=False):
+        root_path = path_info.path
+        seen_paths = set()
+
+        dir_ids = [self._ids_cache["ids"].copy()]
+        while dir_ids:
+            query_ids = {
+                dir_id: dir_name
+                for dir_id, dir_name in dir_ids.pop().items()
+                if posixpath.commonpath([root_path, dir_name])
```

Comment:
maybe I'm missing something, but what happens if `dir_name` path is shorter than the `root_path` (e.g. cached from the previous run). They will have the common prefix right, but we'll end up querying a directory outside the `root_path`?

Reply:
Can you give an example? I don't use commonprefix, but rather use commonpath so it should actually work fine if I am not missing something

Reply:
sure, I was thinking about the case like this:

```
root_path = `a/b/c`
dir_name = `a/b`
```

it will return `a/b`, right? (which seems to be wrong here)

Reply:
Oh, right! Guess we have to add a condition the commonpath is the rootpath.